lib/strategies/webapp-strategy.js

```diff
@@ -207,6 +208,36 @@ function handleAuthorization(req, options, strategy) {
 		options.successRedirect = req.url;
 	}
 
+	if (typeof(strategy.getRefreshToken) === "function") {
+		return obtainTokensWithRefreshToken(req, options, strategy).catch(function() {
+			redirectForAuthorization(req, options, strategy);
+		});
+	}
+	redirectForAuthorization(req, options, strategy);
+}
+
+function obtainTokensWithRefreshToken(req, options, strategy) {
+	return Q.resolve().then(function() {
+		return strategy.getRefreshToken(req);
+	}).then(function(refreshToken) {
+		if (!refreshToken) {
+			throw new Error("Invalid refresh-token");
```

Comment:
throwing an error is misleading since this is legal to not return a refresh token

Reply:
@nitzann  Thanks!

I agree. Changed to "no refresh-token" but I'm still throwing that error. The function "obtainTokensWithRefreshToken()" returns a promise which should be rejected in any case that it couldn't obtain tokens. The flow is much simpler that way. If you have any other suggestions let me know.

Regarding logs,
I added a debug log when no-refresh token returned from the function.
If an error occurred in retrieveTokens, the function itself logs the error.